File: java/low-level-design/src/com/educosys/parkinglot/Level.java
package com.educosys.parkinglot;

import java.util.ArrayList;
import java.util.List;

public class Level {
    private int floorNo;
    private List<ParkingSpot> parkingSpots;

    public Level(){}
    
    public Level(int floorNo, int totalSpots, VehicleType type) {
        this.floorNo = floorNo;
        parkingSpots = new ArrayList<>(totalSpots);
        for(int i=0;i<totalSpots;i++)
        {
            parkingSpots.add(new ParkingSpot(i, type));
        }
    }

    public boolean parkVehicle(Vehicle vehicle)
    {
        for(ParkingSpot parkingSpot: parkingSpots)
        {
            if(parkingSpot.isAvailable() && parkingSpot.getType() == vehicle.getType())
            {
                parkingSpot.parkVehicle(vehicle);
                return true;
            }
        }

        return false;
    }

    public boolean unparkVehicle(Vehicle vehicle)
    {
        for(ParkingSpot parkingSpot: parkingSpots)
        {
            if((parkingSpot.getVehicle() == vehicle) && (!parkingSpot.isAvailable()))
            {
                parkingSpot.unparkVehicle();
                return true;
            }
        }
        return false;
    }

    public void displayParking()
    {
        for(ParkingSpot parkingSpot: parkingSpots)
        {
            System.out.println("Spot No:" + parkingSpot.getSpotNo() + "Availability:" + parkingSpot.isAvailable());
        }
    }
}

File: c++/parking-lot-lld/Level.cpp
#include<bits/stdc++.h>
#include "Level.h"
using namespace std;
        
    Level::Level(){}

    Level::Level(int floorNo, int spotCount, VehicleType type)
    {
        this->floorNo = floorNo;
        for(int i=0;i<spotCount;i++){
            parkingSpots.push_back(new ParkingSpot(i, type));
        }
    }

    bool Level::parkVehicle(Vehicle *vehicle)
    {
        for(auto it: parkingSpots)
        {
            if(it->isParkingAvailable() && it->getVehicleType() == vehicle->getType()) {
                it->parkVehicle(vehicle);
                return true;
            }       
        }

        cout<<endl<<"Cannot park :( by Level"<<endl;
        return false;
    }

    bool Level::unparkVehicle(Vehicle *vehicle)
    {
        for(auto it: parkingSpots){
            if(vehicle == it->getParkedVehicle())
            {
                it->unparkVehicle();
                return true;
            }
        }
        return false;
    }

    void Level::displayLevels(){
        cout<<"Level:"<<floorNo<<endl;
        for(int i=0;i<parkingSpots.size();i++)
        {
            parkingSpots[i]->displayParking();
        }
    }